File: solutions/1072. Flip Columns For Maximum Number of Equal Rows/1072.py
# Time complexity: O(mn)
# Space complexity: O(mn)
class Solution:
  def maxEqualRowsAfterFlips(self, matrix: List[List[int]]) -> int:
    patternFrequency = {}

    for row in matrix:
      rowPattern = "".join(
        "T" if num == row[0] else "F" for num in row
      )

      patternFrequency[rowPattern] = (
        patternFrequency.get(rowPattern, 0) + 1
      )

    return max(patternFrequency.values(), default=0)

File: solutions/1072. Flip Columns For Maximum Number of Equal Rows/1072.cpp
// Time complexity: O(mn)
// Space complexity: O(mn)
class Solution {
 public:
  int maxEqualRowsAfterFlips(const vector<vector<int>>& matrix) {
    int ans = 0;
    unordered_map<string, int> patternFrequency;

    for (auto& row : matrix) {
      string patternBuilder = "";

      for (int col = 0; col < row.size(); col++)
        if (row[0] == row[col])
          patternBuilder += "T";
        else
          patternBuilder += "F";
      patternFrequency[patternBuilder]++;
    }

    for (auto& entry : patternFrequency)
      ans = max(entry.second, ans);

    return ans;
  }
};